core/capabilities/launcher.go

```diff
@@ -0,0 +1,525 @@
+						workflowDON = *toDONInfo(d)
+						w.lggr.Debug("Workflow DON identified: %+v", workflowDON)
+					} else {
+						w.lggr.Errorf("Configuration error: node %s belongs to more than one workflowDON", pid)
+					}
+				}
+
+				capabilityDONs = append(capabilityDONs, *toDONInfo(d))
+			}
+		}
+	}
+
+	w.localNode = capabilities.Node{
+		PeerID:         &pid,
+		WorkflowDON:    workflowDON,
+		CapabilityDONs: capabilityDONs,
+	}
+}
+
+func (w *launcher) Launch(ctx context.Context, state registrysyncer.State) error {
+	w.updateLocalNode(state)
+
+	// Let's start by updating the list of Peers
+	// We do this by creating a new entry for each node belonging
+	// to a public DON.
+	// We also add the hardcoded peers determined by the NetworkSetup.
+	allPeers := make(map[ragetypes.PeerID]p2ptypes.StreamConfig)
+
+	publicDONs := []kcr.CapabilitiesRegistryDONInfo{}
+	for _, d := range state.IDsToDONs {
+		if !d.IsPublic {
+			continue
+		}
+
+		publicDONs = append(publicDONs, d)
+
+		for _, nid := range d.NodeP2PIds {
+			allPeers[nid] = defaultStreamConfig
+		}
+	}
+
+	// TODO: be a bit smarter about who we connect to; we should ideally only
+	// be connecting to peers when we need to.
+	// https://smartcontract-it.atlassian.net/browse/KS-330
+	err := w.peerWrapper.GetPeer().UpdateConnections(allPeers)
+	if err != nil {
+		return fmt.Errorf("failed to update peer connections: %w", err)
+	}
+
+	// Next, we need to split the DONs into the following:
+	// - workflow DONs the current node is a part of.
+	// These will need remote shims to all remote capabilities on other DONs.
+	//
+	// We'll also construct a set to record what DONs the current node is a part of,
+	// regardless of any modifiers (public/acceptsWorkflows etc).
+	myID := w.peerWrapper.GetPeer().ID()
+	myWorkflowDONs := []kcr.CapabilitiesRegistryDONInfo{}
+	remoteWorkflowDONs := []kcr.CapabilitiesRegistryDONInfo{}
+	myDONs := map[uint32]bool{}
+	for _, d := range state.IDsToDONs {
+		for _, peerID := range d.NodeP2PIds {
+			if peerID == myID {
+				myDONs[d.Id] = true
+			}
+		}
+
+		if d.AcceptsWorkflows {
+			if myDONs[d.Id] {
+				myWorkflowDONs = append(myWorkflowDONs, d)
+			} else {
+				remoteWorkflowDONs = append(remoteWorkflowDONs, d)
+			}
+		}
+	}
+
+	// - remote capability DONs (with IsPublic = true) the current node is a part of.
+	// These need server-side shims.
+	myCapabilityDONs := []kcr.CapabilitiesRegistryDONInfo{}
+	remoteCapabilityDONs := []kcr.CapabilitiesRegistryDONInfo{}
+	for _, d := range publicDONs {
+		if len(d.CapabilityConfigurations) > 0 {
+			if myDONs[d.Id] {
+				myCapabilityDONs = append(myCapabilityDONs, d)
+			} else {
+				remoteCapabilityDONs = append(remoteCapabilityDONs, d)
+			}
+		}
+	}
+
+	// Now, if my node is a workflow DON, let's setup any shims
+	// to external capabilities.
+	if len(myWorkflowDONs) > 0 {
+		myDON := myWorkflowDONs[0]
+
+		// TODO: this is a bit nasty; figure out how best to handle this.
+		if len(myWorkflowDONs) > 1 {
```

Comment:
This is no longer necessary because it is ensured onchain, right?

Reply:
I think we do still need some variant of this check (an error since it shouldn't happen) because the DONs are returned as a list

Reply:
I've reworded it though to make it clear that this is an invariant violation